staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go

```diff
@@ -225,10 +255,78 @@ func (r *proxyHandler) updateAPIService(apiService *apiregistrationv1api.APIServ
 		serviceNamespace: apiService.Spec.Service.Namespace,
 		servicePort:      *apiService.Spec.Service.Port,
 		serviceAvailable: apiregistrationv1apihelper.IsAPIServiceConditionTrue(apiService, apiregistrationv1api.Available),
+		defaultSNI:       transportConfig.TLS.ServerName,
+	}
+
+	r.updateHandlingInfo(&newInfo)
+}
+
+func (r *proxyHandler) updateHandlingInfo(handlingInfo *proxyHandlingInfo) {
+	handlingInfo.proxyRoundTripper, handlingInfo.transportBuildingError = transport.New(handlingInfo.transportConfig)
+	if handlingInfo.transportBuildingError != nil {
+		klog.Warning(handlingInfo.transportBuildingError.Error())
 	}
-	newInfo.proxyRoundTripper, newInfo.transportBuildingError = transport.New(newInfo.transportConfig)
-	if newInfo.transportBuildingError != nil {
-		klog.Warning(newInfo.transportBuildingError.Error())
+	r.handlingInfo.Store(*handlingInfo)
+}
+
+func newFallbackRoundTripper(proxyHandler *proxyHandler, roundTripper http.RoundTripper) *fallbackRoundTripper {
+	return &fallbackRoundTripper{
+		proxyHandler:         proxyHandler,
+		originalRoundTripper: roundTripper,
 	}
-	r.handlingInfo.Store(newInfo)
+}
+
+var _ utilnet.RoundTripperWrapper = &fallbackRoundTripper{}
+
+type fallbackRoundTripper struct {
+	proxyHandler         *proxyHandler
+	originalRoundTripper http.RoundTripper
+}
+
+func (f *fallbackRoundTripper) WrappedRoundTripper() http.RoundTripper {
+	return f.originalRoundTripper
+}
+
+func (f *fallbackRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	value := f.proxyHandler.handlingInfo.Load()
+	if value == nil {
+		return f.originalRoundTripper.RoundTrip(req)
+	}
+	handlingInfo := value.(proxyHandlingInfo)
+
+	// return early if the request is already using the default SNI
+	if handlingInfo.transportConfig == nil || handlingInfo.transportConfig.TLS.ServerName == handlingInfo.defaultSNI {
+		return f.originalRoundTripper.RoundTrip(req)
+	}
+
+	// clone the request for use with retry
+	clone := utilnet.CloneRequest(req)
+
+	resp, err := f.originalRoundTripper.RoundTrip(req)
+	// return error if the error is not an x509 HostnameError
+	if err != nil {
+		verificationErr, isVerification := err.(*tls.CertificateVerificationError)
+		if !isVerification {
+			return resp, err
+		}
+		if _, isHostname := verificationErr.Unwrap().(x509.HostnameError); !isHostname {
+			return resp, err
+		}
+	}
+
+	if resp != nil {
+		return resp, err
+	}
+
+	// skip external SNI resolution, and always use the default SNI
+	handlingInfo.fallbackToDefaultSNI = true
```

Comment:
should this fallback behavior be deprecated at some point? if so, should i log a warning here about the deprecation?

Reply:
I see no reason to break existing users, there is nothing inherently wrong or insecure about them using the other server name.